### PR TITLE
feat(DWS): add workload plan stage resource support

### PR DESCRIPTION
**Pull request by zhangtao505**
**What this PR does / why we need it**:
 add workload plan stage resource support

**Special notes for your reviewer**:

**Release note**:
 add workload plan stage resource support

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

make testacc TEST=./huaweicloud/services/acceptance/dws TESTARGS='-run TestAccResourceWorkLoadPlanStage'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dws -v -run TestAccResourceWorkLoadPlanStage -timeout 360m -parallel 4
=== RUN   TestAccResourceWorkLoadPlanStage_basic
=== PAUSE TestAccResourceWorkLoadPlanStage_basic
=== CONT  TestAccResourceWorkLoadPlanStage_basic
--- PASS: TestAccResourceWorkLoadPlanStage_basic (1306.47s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dws       1306.524s

### CODE REVIEW

File: docs/resources/dws_workload_plan_stage.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+```hcl
+variable "cluster_id" {}
+variable "plan_id" {}
+variable "stage_name" {}
+variable "pool_name" {}
+
+resource "huaweicloud_dws_workload_plan_stage" "test" {
+  cluster_id = var.cluster_id
+  plan_id    = var.cluster_id
```
Comment (Lance52259) [inline docs/resources/dws_workload_plan_stage.md]:
```suggestion
  plan_id    = var.plan_id
```

File: docs/resources/dws_workload_plan_stage.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+* `queue_list` - (Required, List, ForceNew) Specifies the workload queue configurations.
+  Changing this creates a new resource.
+  The [queue_list](#block--queue_list) structure is documented below.
+
+* `month` - (Optional, String, ForceNew) Specifies the execution months of the workload plan. Default to all months.
+  Changing this creates a new resource.
+
+* `day` - (Optional, String, ForceNew) Specifies the execution days of the workload plan. Default to all days.
+  Changing this creates a new resource.
```
Comment (Lance52259) [inline docs/resources/dws_workload_plan_stage.md]:
```
e.g. **...**
```
please provider the examples to teach users how to use them.

File: docs/resources/dws_workload_plan_stage.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID.
+
+* `next_valid_time` - Next execution time.
+
+## Import
+
+The workload plan stage can be imported using `cluster_id` and `plan_id` and `name`, separated by a slash, e.g.
```
Comment (Lance52259) [inline docs/resources/dws_workload_plan_stage.md]:
```suggestion
The workload plan stage can be imported using `cluster_id` and `plan_id` and `name`, separated by slashes, e.g.
```

File: docs/resources/dws_workload_plan_stage.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+* `resource_description` - (Optional, String, ForceNew) Specifies the description of the resource attribute for
+  the workload queue. Changing this creates a new resource.
+
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID.
+
+* `next_valid_time` - Next execution time.
```
Comment (Lance52259) [inline docs/resources/dws_workload_plan_stage.md]: execution time? not validation time?

File: docs/resources/dws_workload_plan_stage.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  Changing this creates a new resource.
+
+<a name="block--queue_list--configuration"></a>
+The `configuration` block supports:
+
+* `resource_name` - (Required, String, ForceNew) Specifies the resource name to be configured for the workload queue.
+  Changing this creates a new resource. Value options:
+
+  + **cpu**: Percentage of CPU time that can be used by users associated with the current workload queue to execute jobs.
+    The value is an integer ranging from `1` to `99`.
```
Comment (Lance52259) [inline docs/resources/dws_workload_plan_stage.md]: This description is belong to the parameter `resource_value`

File: huaweicloud/services/acceptance/dws/resource_huaweicloud_dws_workload_plan_stage_test.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	getPath = strings.ReplaceAll(getPath, "{project_id}", client.ProjectID)
+	getPath = strings.ReplaceAll(getPath, "{cluster_id}", state.Primary.Attributes["cluster_id"])
+	getPath = strings.ReplaceAll(getPath, "{plan_id}", state.Primary.Attributes["plan_id"])
+	getOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+
+	resp, err := client.Request("GET", getPath, &getOpt)
+	if err != nil {
+		if _, ok := err.(golangsdk.ErrDefault403); ok {
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/dws/resource_huaweicloud_dws_workload_plan_stage_test.go]: Why do we return authentication errors (error code 403) as errors of the resource does not exist (error code 404)?

File: huaweicloud/services/acceptance/dws/resource_huaweicloud_dws_workload_plan_stage_test.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+	respBody, err := utils.FlattenResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	respErrCode := utils.PathSearch("error_code", respBody, "")
+	if respErrCode != "" {
+		return nil, fmt.Errorf("error retrieving DWS workload plan stage: %s", utils.PathSearch("workload_res_str", respBody, nil))
+	}
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/dws/resource_huaweicloud_dws_workload_plan_stage_test.go]: we cannot guarantee that all error codes have a corresponding `workload_res_str`. please make sure that nil is not returned in this case.

File: huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+func flattenPoolConfiguration(queue interface{}) []interface{} {
+	poolCon := utils.PathSearch("queue_resources", queue, make([]interface{}, 0)).([]interface{})
+	if len(poolCon) == 0 {
+		return nil
+	}
+	rst := make([]interface{}, len(poolCon))
+	for i, queue := range poolCon {
+		rst[i] = map[string]interface{}{
+			"resource_name":        utils.PathSearch("resource_name", queue, ""),
+			"resource_value":       utils.PathSearch("resource_value", queue, ""),
```
Comment (jinyangyang222) [inline huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go:299]: resource_value type is int

File: huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+	}
+
+	return params
+}
+
+func resourceWorkLoadPlanStageRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg     = meta.(*config.Config)
+		region  = cfg.GetRegion(d)
+		httpUrl = "v2/{project_id}/clusters/{cluster_id}/workload/plans/{plan_id}"
```
Comment (jinyangyang222) [inline huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go:218]: why not use the API for querying stage details

File: docs/resources/dws_workload_plan_stage.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+variable "plan_id" {}
+variable "stage_name" {}
+variable "pool_name" {}
+
+resource "huaweicloud_dws_workload_plan_stage" "test" {
+  cluster_id = var.cluster_id
+  plan_id    = var.cluster_id
+  name       = var.stage_nme
+  start_time = "01:00:00"
+  end_time   = "00:00:00"
```
[the remark this answers is not in the document]
Comment (zhangtao505) [reply docs/resources/dws_workload_plan_stage.md:22]: done

File: docs/resources/dws_workload_plan_stage.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  queue_list {
+    name =  var.pool_name
```
[the remark this answers is not in the document]
Comment (zhangtao505) [reply docs/resources/dws_workload_plan_stage.md]: done

File: docs/resources/dws_workload_plan_stage.md (hunk 1 above)
Comment (zhangtao505) [reply docs/resources/dws_workload_plan_stage.md]: done

File: docs/resources/dws_workload_plan_stage.md (hunk 2 above)
Comment (zhangtao505) [reply docs/resources/dws_workload_plan_stage.md]: done

File: docs/resources/dws_workload_plan_stage.md (hunk 3 above)
Comment (zhangtao505) [reply docs/resources/dws_workload_plan_stage.md]: done

File: docs/resources/dws_workload_plan_stage.md (hunk 4 above)
Comment (zhangtao505) [reply docs/resources/dws_workload_plan_stage.md]: done

File: docs/resources/dws_workload_plan_stage.md (hunk 5 above)
Comment (zhangtao505) [reply docs/resources/dws_workload_plan_stage.md]: done

File: huaweicloud/services/acceptance/dws/resource_huaweicloud_dws_workload_plan_stage_test.go (hunk 7 above)
Comment (zhangtao505) [reply huaweicloud/services/acceptance/dws/resource_huaweicloud_dws_workload_plan_stage_test.go]: done

File: huaweicloud/services/acceptance/dws/resource_huaweicloud_dws_workload_plan_stage_test.go (hunk 6 above)
Comment (zhangtao505) [reply huaweicloud/services/acceptance/dws/resource_huaweicloud_dws_workload_plan_stage_test.go]: done

File: huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go (hunk 8 above)
Comment (zhangtao505) [reply huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go]: done

File: huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go (hunk 9 above)
Comment (zhangtao505) [reply huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go:220]: the plan details API contains all stage information and does not require further queries

File: huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+		return diag.Errorf("error creating DWS workload plan stage: %s", err)
+	}
+	respBody, err := utils.FlattenResponse(resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	respErrCode := utils.PathSearch("error_code", respBody, "")
+	if respErrCode != "" {
+		return diag.Errorf("error creating DWS workload plan stage: %s", utils.PathSearch("workload_res_str", respBody, nil))
+	}
```
[the remark this answers is not in the document]
Comment (zhangtao505) [reply huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go:162]: the ID set with detail information

File: huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/chnsz/golangsdk"
+
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+var defaultErrMsg = "The resource does not exist or is illegal, please contact technical support for assistance"
```
Comment (Lance52259) [inline huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go]:
```suggestion
var defaultErrMsg = "the resource is illegal, please contact technical support"
```
It's not possible to return a `resource not found` error at the reference codes where you return this error code, whether create or update.

Comment (zhangtao505) [reply huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go]: done

File: huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go (hunk 9 above)
Comment (zhangtao505) [reply huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go]: done

File: huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (zhangtao505) [reply huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go:163]: done

**Comment by zhangtao505** [comment]
make testacc TEST=./huaweicloud/services/acceptance/dws TESTARGS='-run TestAccResourceWorkLoadPlanStage'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dws -v -run TestAccResourceWorkLoadPlanStage -timeout 360m -parallel 4
=== RUN   TestAccResourceWorkLoadPlanStage_basic
=== PAUSE TestAccResourceWorkLoadPlanStage_basic
=== CONT  TestAccResourceWorkLoadPlanStage_basic
--- PASS: TestAccResourceWorkLoadPlanStage_basic (1306.90s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dws       1306.944s

**Pull request merged**